File: jbake-dist/src/smoke-test/java/org/jbake/ProjectWebsiteTest.java
package org.jbake;

import org.apache.commons.vfs2.util.Os;
import org.eclipse.jgit.api.CloneCommand;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.junit.Assume;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.File;
import java.io.IOException;

import static org.assertj.core.api.Assertions.assertThat;

public class ProjectWebsiteTest {


    private static final String WEBSITE_REPO_URL = "https://github.com/jbake-org/jbake.org.git";

    @Rule
    public TemporaryFolder folder = new TemporaryFolder();
    private File projectFolder;
    private File templateFolder;
    private File outputFolder;
    private String jbakeExecutable;
    private BinaryRunner runner;

    @Before
    public void setup() throws IOException, GitAPIException {
        Assume.assumeTrue("JDK 7 is not supported for this test", !isJava7());
        if ( Os.isFamily(Os.OS_FAMILY_WINDOWS) ) {
            jbakeExecutable = new File("build\\install\\jbake\\bin\\jbake.bat").getAbsolutePath();
        }
        else {
            jbakeExecutable = new File("build/install/jbake/bin/jbake").getAbsolutePath();
        }
        projectFolder = folder.newFolder("project");
        templateFolder = new File(projectFolder, "templates");
        outputFolder = new File(projectFolder, "output");

        runner = new BinaryRunner(projectFolder);
        cloneJbakeWebsite();

    }

    private boolean isJava7() {
        return System.getProperty("java.specification.version").equals("1.7");
    }

    private void cloneJbakeWebsite() throws GitAPIException {
        CloneCommand cmd = Git.cloneRepository();
        cmd.setBare(false);
        cmd.setBranch("master");
        cmd.setRemote("origin");
        cmd.setURI(WEBSITE_REPO_URL);
        cmd.setDirectory(projectFolder);

        Git git = cmd.call();

        assertThat(new File(projectFolder,"README.md").exists()).isTrue();
    }

    @Test
    public void shouldBakeWebsite() throws IOException, InterruptedException {
        Process process = runner.runWithArguments(jbakeExecutable,"-b");
        assertThat(process.exitValue()).isEqualTo(0);
        assertThat(new File(outputFolder, "index.html")).exists();
        process.destroy();
    }

}

File: jbake-dist/src/smoke-test/kotlin/org/jbake/ProjectWebsiteTest.kt
package org.jbake

import io.kotest.assertions.withClue
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.file.shouldExist
import io.kotest.matchers.shouldBe
import org.eclipse.jgit.api.Git
import kotlin.io.path.createTempDirectory

class ProjectWebsiteTest : StringSpec({

    "should bake JBake website" {
        val tempDir = createTempDirectory(buildOutputDir, "jbake-website-test-").toFile()
        try {
            val projectDir = tempDir.resolve("project")
            val outputDir = projectDir.resolve("output")

            // Clone JBake website repository
            Git.cloneRepository().setURI(WEBSITE_REPO_URL).setDirectory(projectDir).setBare(false).setBranch("master").setRemote("origin").call()
            projectDir.resolve("README.md").shouldExist()

            // Bake the website (with verbose logging to debug failures)
            val runner = BinaryRunner(projectDir)
            val process = runner.runWithArguments(BinaryRunner.jbakeExecutableRelative.absolutePath, "-vvv", "-b")
            withClue("\n" +
                "========= JBake process output: =========\n\n${runner.processOutput}\n" +
                "=========================================\nProcess exit code:") {
                process.exitValue() shouldBe 0
            }
            outputDir.resolve("index.html").shouldExist()
            process.destroy()
        }
        finally {
            tempDir.deleteRecursively()
        }
    }
})

private const val WEBSITE_REPO_URL = "https://github.com/jbake-org/jbake.org.git"
